src/deephaven_ib/_tws/tws_client.py

```diff
@@ -0,0 +1,926 @@
+             dht.int32, dht.string, dht.float64])
+
+        table_writers["orders_completed"] = TableWriter(
+            [*logger_contract.names(), *logger_order.names(), *logger_order_state.names()],
+            [*logger_contract.types(), *logger_order.types(), *logger_order_state.types()])
+
+        table_writers["orders_exec_details"] = TableWriter(
+            ["RequestId", *logger_contract.names(renames={"Exchange": "ContractExchange"}),
+             *logger_execution.names(renames={"Exchange": "ExecutionExchange"})],
+            [dht.int32, *logger_contract.types(), *logger_execution.types()])
+
+        table_writers["orders_exec_commission_report"] = TableWriter(
+            [*logger_commission_report.names()],
+            [*logger_commission_report.types()])
+
+        ####
+        # End
+        ####
+
+        return table_writers
+
+    ####################################################################################################################
+    ####################################################################################################################
+    ## Connect / Disconnect / Subscribe
+    ####################################################################################################################
+    ####################################################################################################################
+
+    def connect(self, host: str, port: int, client_id: int) -> None:
+        """Connect to an IB TWS session.  Raises an exception if already connected.
+
+        Args:
+            host (str): The host name or IP address of the machine where TWS is running. Leave blank to connect to the local host.
+            port (int): TWS port, specified in TWS on the Configure>API>Socket Port field.
+                By default production trading uses port 7496 and paper trading uses port 7497.
+            client_id (int): A number used to identify this client connection.
+                All orders placed/modified from this client will be associated with this client identifier.
+
+                Note: Each client MUST connect with a unique clientId.
+
+        Returns:
+              None
+
+        Raises:
+              Exception
+        """
+
+        if self.isConnected():
+            raise Exception("IbTwsClient is already connected.")
+
+        self.contract_registry = ContractRegistry(self)
+        self.order_id_queue = OrderIdEventQueue(self)
+        self._registered_market_rules = set()
+        self._realtime_bar_sizes = {}
+        self.news_providers = []
+        self._accounts_managed = set()
+
+        EClient.connect(self, host, port, client_id)
+
+        # wait for the client to connect to avoid a race condition
+        time.sleep(1)
+
+        self._thread = Thread(target=self.run)
+        self._thread.start()
+        setattr(self, "ib_thread", self._thread)
+
+        # wait for the client to connect to avoid a race condition
+        time.sleep(1)
+
```

Comment:
not sure using sleep() here is deterministic, maybe a while loop with isConnected() works better?

Reply:
I initially tried that.  The race condition also involves when `isConnected()` gets updated.  It is a real mess.  Cost me 2 days to get to these sleeps.  They suck.